src/ripple/app/misc/NetworkOPs.cpp

```diff
@@ -619,6 +639,63 @@ class NetworkOPsImp final
     std::vector <TransactionStatus> mTransactions;
 
     StateAccounting accounting_ {};
+
+private:
+    struct Stats
+    {
+        template <class Handler>
+        Stats (Handler const& handler, beast::insight::Collector::ptr const& collector)
+            : hook (collector->make_hook (handler))
+            , disconnected_duration (collector->make_gauge("State_Accounting","Disconnected_duration"))
+            , connected_duration (collector->make_gauge("State_Accounting","Connected_duration"))
+            , syncing_duration(collector->make_gauge("State_Accounting", "Syncing_duration"))
+            , tracking_duration (collector->make_gauge("State_Accounting", "Tracking_duration"))
+            , full_duration (collector-> make_gauge("State_Accounting", "Full_duration"))
+            , disconnected_trasitions (collector->make_gauge("State_Accounting","Disconnected_trasitions"))
+            , connected_trasitions (collector->make_gauge("State_Accounting","Connected_trasitions"))
+            , syncing_trasitions(collector->make_gauge("State_Accounting", "Syncing_trasitions"))
+            , tracking_trasitions (collector->make_gauge("State_Accounting", "Tracking_trasitions"))
+            , full_trasitions (collector-> make_gauge("State_Accounting", "Full_trasitions"))
+            { }
+
+        beast::insight::Hook hook;
+        beast::insight::Gauge disconnected_duration;
+        beast::insight::Gauge connected_duration;
+        beast::insight::Gauge syncing_duration;
+        beast::insight::Gauge tracking_duration;
+        beast::insight::Gauge full_duration;
+
+        beast::insight::Gauge disconnected_trasitions;
+        beast::insight::Gauge connected_trasitions;
+        beast::insight::Gauge syncing_trasitions;
+        beast::insight::Gauge tracking_trasitions;
+        beast::insight::Gauge full_trasitions;
+    };
+
+    Stats m_stats;
+
+private:
+    void collect_metrics()
+    {   
+        auto counters = accounting_.getCounters();
+        auto start = accounting_.getStart();
+        auto mode = accounting_.getMode();
+
+        auto const current = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::system_clock::now() - start);
+        counters[static_cast<std::size_t>(mode)].dur += current;
+
+        m_stats.disconnected_duration.set(counters[static_cast<std::size_t>(OperatingMode::DISCONNECTED)].dur.count());
```

Comment:
does `m_stats` or `accounting_` need to be under a lock? Seems like they might need to be, although I'm not sure when this callback is triggered (`beast::insight::Hook`?)

Reply:
This call back is triggered once a second via the StatsDCollector. See here:

https://github.com/ripple/rippled/blob/90d9ca901d06b9da1653526c234bc4b00e27070b/src/ripple/beast/insight/impl/StatsDCollector.cpp#L418

From what I can tell, I don't think they need to be under a lock.

Reply:
So currently `accounting_` is safe, since it protects it's members, and `m_stats` is only modified here. But you don't control when it's called/from what thread, so it might be better to protect `m_stats` as well just to protect against future changes.